src/server/controllers/partner-details.controller.js

```diff
@@ -0,0 +1,271 @@
+      logEmitter.emit(
+        "functionSuccessWith",
+        "partner-details.controller",
+        "partnerDetailsContinue",
+        `validatorErrors: ${JSON.stringify(
+          controllerResponse.validatorErrors
+        )}. redirectRoute: ${controllerResponse.redirectRoute}`
+      );
+      return controllerResponse;
+    }
+    controllerResponse.redirectRoute = `/new/${council}/main-partnership-contact`;
+
+    if (edit) {
+      if (
+        previousAnswers.partners.find(partnerName => {
+          return partnerName === previousAnswers.main_partnership_contact;
+        }) !== undefined
+      ) {
+        controllerResponse.redirectRoute = `/new/${council}/registration-summary`;
+      } else {
+        controllerResponse.redirectRoute = `/new/${council}/main-partnership-contact?edit=main-partnership-contact`;
+      }
+    }
+
+    logEmitter.emit(
+      "functionSuccess",
+      "partner-details.controller",
+      "partnerDetailsContinue"
+    );
+    return controllerResponse;
+  } catch (err) {
+    logEmitter.emit(
+      "functionFail",
+      "partner-details.controller",
+      "partnerDetailsContinue",
+      err
+    );
+    throw err;
+  }
+};
+
+/**
+ * Returns an object containing validator errors (if present), the redirect route (e.g. the next page),
+ * the new value of the cumulative answers object, and the new value of the switches object.
+ *
+ * @param {string} currentPage The 'originator' page that the user has come from
+ * @param {object} previousAnswers An object containing every past answer that has been given by the user
+ * @param {object} newAnswers An object containing new answers from the current page
+ * @param {string} council local_council_url allowing to put the user back on the path by redirecting to /new/council
+ * @param {boolean} edit Flag indicating if request was submitted in edit mode
+ *
+ * @returns {object} Values for the router to store/update in the session and the page to redirect to.
+ */
+const partnerDetailsSave = (
+  currentPage,
+  previousAnswers,
+  newAnswers,
+  council,
+  edit
+) => {
+  const controllerResponse = {
+    validatorErrors: {},
+    redirectRoute: null,
+    cumulativeFullAnswers: {},
+    addressLookups: {},
+    switches: {}
+  };
+  logEmitter.emit(
+    "functionCall",
+    "partner-details.controller",
+    "partnerDetailsSave"
+  );
+  try {
+    controllerResponse.cumulativeFullAnswers = previousAnswers;
+    controllerResponse.validatorErrors = Object.assign(
+      {},
+      validate(currentPage, newAnswers).errors
+    );
+
+    const validationErrorsKeys = Object.keys(
+      controllerResponse.validatorErrors
+    );
+    if (validationErrorsKeys.length > 0) {
+      // ignore if it's only the partners array
+      if (
+        validationErrorsKeys[0] === "partners" &&
+        validationErrorsKeys.length === 1
+      ) {
+        controllerResponse.validatorErrors = {};
+      } else {
```

Comment:
I really don't like this? Is there an alternative?

Reply:
Yes we can extract the partner_name from newAnswers and validate just that one field explicitly.